ducktape/mark/_mark.py

```diff
@@ -0,0 +1,359 @@
+# Copyright 2015 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from ducktape.tests.test import TestContext
+from ducktape.errors import DucktapeError
+
+import functools
+import itertools
+
+
+class Mark(object):
+    """Common base class for "marks" which may be applied to test functions/methods."""
+
+    @staticmethod
+    def mark(fun, mark):
+        """Attach a tag indicating that fun has been marked with the given mark
+
+        Marking fun updates it with two attributes:
+
+        - marks:      a list of mark objects applied to the function. These may be strings or objects subclassing Mark
+                      we use a list because in some cases, it is useful to preserve ordering.
+        - mark_names: a set of names of marks applied to the function
+        """
+        # Update fun.marks
+        if hasattr(fun, "marks"):
```

Comment:
You're seeing the effect of the hooks that Python provides for intercepting `getattr` and `setattr` by overriding special methods on the object: https://docs.python.org/2/reference/datamodel.html#object.__getattr__

This is partially why I'm asking -- the two paths you're using have differences, so you could end up with inconsistencies. I think the way you're using it here ends up fine since `__getattr__` is only called if the attribute wasn't found via normal lookups, which will include checking the `__dict__`, but the inconsistency is odd since you can also read from `__dict__`.